File: agent/visual_analyser/kafka_consumer.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"os/signal"
	"strings"
	"sync"
	"syscall"
	"time"

	"github.com/google/uuid"
	"github.com/segmentio/kafka-go"
	"jarvis/agent/schemas/group"
)

// KafkaConsumer consumes events from the group communication layer
type KafkaConsumer struct {
	reader         *kafka.Reader
	producer       *kafka.Writer
	groupID        string
	agentID        string
	brokers        []string
	mu             sync.Mutex
	running        bool
	visionService  *VisionService
}

// NewKafkaConsumer creates a new Kafka consumer for the visual analyser
func NewKafkaConsumer(brokers []string, groupID, agentID string, visionService *VisionService) (*KafkaConsumer, error) {
	reader := kafka.NewReader(kafka.ReaderConfig{
		Brokers:        brokers,
		Topic:          group.EventsTopic(groupID),
		GroupID:        fmt.Sprintf("%s-group", agentID),
		StartOffset:    kafka.LastOffset,
		CommitInterval: time.Second,
		SessionTimeout: 10 * time.Second,
		MaxBytes:       1024 * 1024, // 1MB
	})

	producer := &kafka.Writer{
		Addr:     kafka.TCP(brokers...),
		Topic:    group.EventsTopic(groupID),
		Balancer: &kafka.LeastBytes{},
	}

	return &KafkaConsumer{
		reader:        reader,
		producer:      producer,
		groupID:       groupID,
		agentID:       agentID,
		brokers:       brokers,
		running:       false,
		visionService: visionService,
	}, nil
}

// createEnvelope creates a group layer envelope for publishing
func (kc *KafkaConsumer) createEnvelope(eventType string, payload map[string]interface{}, threadID string) *group.Envelope {
	if threadID == "" {
		threadID = uuid.New().String()
	}

	return group.NewEnvelope(
		kc.groupID,
		threadID,
		uuid.New().String(),
		group.Sender{
			ID:    kc.agentID,
			Role:  group.RoleMember,
			Agent: kc.agentID,
		},
		eventType,
		payload,
	)
}

// publishEvent publishes an event to the events topic
func (kc *KafkaConsumer) publishEvent(ctx context.Context, eventType string, payload map[string]interface{}, threadID string) error {
	envelope := kc.createEnvelope(eventType, payload, threadID)

	data, err := json.Marshal(envelope)
	if err != nil {
		return fmt.Errorf("failed to marshal envelope: %v", err)
	}

	message := kafka.Message{
		Key:   []byte(envelope.ThreadID),
		Value: data,
	}

	if err := kc.producer.WriteMessages(ctx, message); err != nil {
		return fmt.Errorf("failed to publish event: %v", err)
	}

	log.Printf("Published %s event: %s", eventType, envelope.MessageID)
	return nil
}

// handleCommand handles a command execution request
func (kc *KafkaConsumer) handleCommand(ctx context.Context, envelope *group.Envelope) error {
	targetAgent, _ := envelope.Payload["target_agent"].(string)
	command, _ := envelope.Payload["command"].(string)
	taskID, _ := envelope.Payload["task_id"].(string)
	if taskID == "" {
		taskID = uuid.New().String()
	}

	if targetAgent != kc.agentID {
		log.Printf("Command not for this agent (%s != %s)", targetAgent, kc.agentID)
		return nil
	}

	log.Printf("Received command: %s", command)

	// Publish task started event
	if err := kc.publishEvent(ctx, group.EventTaskStarted, map[string]interface{}{
		"task_id":       taskID,
		"target_agent":  kc.agentID,
		"status":        "started",
		"thread_id":     envelope.ThreadID,
	}, envelope.ThreadID); err != nil {
		log.Printf("Failed to publish task started event: %v", err)
	}

	// Execute the analysis
	imageURL, _ := envelope.Payload["image_url"].(string)
	imageData, _ := envelope.Payload["image_data"].(string)
	prompt, _ := envelope.Payload["prompt"].(string)

	if imageURL == "" && imageData == "" {
		errMsg := "No image provided"
		log.Printf("Error: %s", errMsg)

		if err := kc.publishEvent(ctx, group.EventTaskFailed, map[string]interface{}{
			"task_id":   taskID,
			"error":     errMsg,
			"status":    "failed",
			"thread_id": envelope.ThreadID,
		}, envelope.ThreadID); err != nil {
			log.Printf("Failed to publish task failed event: %v", err)
		}
		return nil
	}

	// Run vision analysis
	var result string
	var err error

	if imageData != "" {
		result, err = kc.visionService.AnalyzeImage(imageData, prompt)
	} else if imageURL != "" {
		// For URL, fetch it first - simplified to just analyze if we had base64
		result, err = kc.visionService.AnalyzeImage(imageURL, prompt)
	} else {
		err = fmt.Errorf("no image data or URL provided")
	}

	if err != nil {
		log.Printf("Error analyzing image: %v", err)

		if err := kc.publishEvent(ctx, group.EventTaskFailed, map[string]interface{}{
			"task_id":   taskID,
			"error":     err.Error(),
			"status":    "failed",
			"thread_id": envelope.ThreadID,
		}, envelope.ThreadID); err != nil {
			log.Printf("Failed to publish task failed event: %v", err)
		}
		return nil
	}

	// Publish task completed event
	if err := kc.publishEvent(ctx, group.EventTaskCompleted, map[string]interface{}{
		"task_id":   taskID,
		"status":    "completed",
		"result":    result,
		"duration":  0,
		"thread_id": envelope.ThreadID,
	}, envelope.ThreadID); err != nil {
		log.Printf("Failed to publish task completed event: %v", err)
	}

	// Publish result event
	if err := kc.publishEvent(ctx, group.EventResultImageAnalysis, map[string]interface{}{
		"task_id":   taskID,
		"type":      "image_analysis",
		"data":      map[string]interface{}{"analysis": result},
		"success":   true,
		"duration":  0,
		"thread_id": envelope.ThreadID,
	}, envelope.ThreadID); err != nil {
		log.Printf("Failed to publish result event: %v", err)
	}

	return nil
}

// processMessage processes a single Kafka message
func (kc *KafkaConsumer) processMessage(ctx context.Context, msg kafka.Message) error {
	var envelope group.Envelope

	if err := json.Unmarshal(msg.Value, &envelope); err != nil {
		return fmt.Errorf("failed to unmarshal envelope: %v", err)
	}

	// Validate envelope
	if err := group.ValidateEnvelope(&envelope); err != nil {
		return fmt.Errorf("invalid envelope: %v", err)
	}

	// Check group matches
	if envelope.GroupID != kc.groupID {
		return fmt.Errorf("envelope group_id mismatch: expected %s, got %s", kc.groupID, envelope.GroupID)
	}

	// Route by event type
	switch envelope.Type {
	case group.EventCommandRun:
		return kc.handleCommand(ctx, &envelope)
	default:
		log.Printf("Ignoring event type: %s", envelope.Type)
		return nil
	}
}

// Start starts consuming messages from Kafka
func (kc *KafkaConsumer) Start(ctx context.Context) error {
	kc.mu.Lock()
	if kc.running {
		kc.mu.Unlock()
		return fmt.Errorf("consumer already running")
	}
	kc.running = true
	kc.mu.Unlock()

	log.Printf("Starting Kafka consumer for %s in group %s", kc.agentID, kc.groupID)

	// Setup signal handling
	sigChan := make(chan os.Signal, 1)
	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM)

	for {
		select {
		case <-ctx.Done():
			return ctx.Err()
		case <-sigChan:
			log.Println("Received shutdown signal, stopping consumer...")
			return kc.Stop()
		default:
		}

		// Read message with timeout
		msgCtx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
		msg, err := kc.reader.ReadMessage(msgCtx)
		cancel()

		if err != nil {
			if err == context.DeadlineExceeded {
				continue
			}
			log.Printf("Error reading message: %v", err)
			continue
		}

		// Process message
		if err := kc.processMessage(ctx, msg); err != nil {
			log.Printf("Error processing message: %v", err)
		}
	}
}

// Stop stops the consumer
func (kc *KafkaConsumer) Stop() error {
	kc.mu.Lock()
	defer kc.mu.Unlock()

	if !kc.running {
		return fmt.Errorf("consumer not running")
	}

	kc.running = false

	if err := kc.reader.Close(); err != nil {
		return fmt.Errorf("failed to close reader: %v", err)
	}

	if err := kc.producer.Close(); err != nil {
		return fmt.Errorf("failed to close producer: %v", err)
	}

	log.Println("Consumer stopped")
	return nil
}

// IsRunning returns whether the consumer is running
func (kc *KafkaConsumer) IsRunning() bool {
	kc.mu.Lock()
	defer kc.mu.Unlock()
	return kc.running
}

// parseBrokers parses broker string and returns slice of broker addresses
func parseBrokers(brokerString string) []string {
	if brokerString == "" {
		brokerString = "localhost:9092"
	}

	brokers := strings.Split(brokerString, ",")
	for i, b := range brokers {
		brokers[i] = strings.TrimSpace(b)
	}

	return brokers
}

// StartKafkaConsumer starts the Kafka consumer in a goroutine if enabled
func StartKafkaConsumer(visionService *VisionService) (*KafkaConsumer, error) {
	if os.Getenv("ENABLE_GROUP_LAYER") != "true" {
		log.Println("Group layer not enabled, skipping Kafka consumer")
		return nil, nil
	}

	groupID := os.Getenv("GROUP_ID")
	if groupID == "" {
		groupID = "jarvis-main"
	}

	agentID := os.Getenv("AGENT_ID")
	if agentID == "" {
		agentID = "visual-analyser"
	}

	brokerString := os.Getenv("KAFKA_BROKERS")
	brokers := parseBrokers(brokerString)

	log.Printf("Initializing Kafka consumer for %s in group %s", agentID, groupID)

	consumer, err := NewKafkaConsumer(brokers, groupID, agentID, visionService)
	if err != nil {
		return nil, fmt.Errorf("failed to create consumer: %v", err)
	}

	// Start consumer in a goroutine
	go func() {
		ctx := context.Background()
		if err := consumer.Start(ctx); err != nil {
			log.Printf("Consumer exited with error: %v", err)
		}
	}()

	time.Sleep(100 * time.Millisecond)
	if !consumer.IsRunning() {
		return nil, fmt.Errorf("failed to start consumer")
	}

	log.Println("Kafka consumer started successfully")
	return consumer, nil
}

File: agent/web_fetcher/kafka_consumer.py
"""
Kafka consumer for the Web Fetcher that listens to group communication events.
Integrates the web fetcher with the Jarvis group communication layer.
"""

import json
import logging
import os
import sys
from datetime import datetime
from typing import Any, Dict, Optional
from uuid import uuid4

from kafka import KafkaConsumer, KafkaProducer
from kafka.errors import KafkaError

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)


class GroupConsumerException(Exception):
    """Custom exception for group consumer errors"""
    pass


class WebFetcherKafkaConsumer:
    """
    Kafka consumer for the Web Fetcher that participates in group communication.
    """

    def __init__(
        self,
        group_id: str = "jarvis-main",
        agent_id: str = "web-fetcher",
        broker_url: str = "localhost:9092",
    ):
        """
        Initialize the Kafka consumer for group communication.

        Args:
            group_id: The group ID for the orchestrator
            agent_id: This agent's unique ID
            broker_url: Kafka broker URL
        """
        self.group_id = group_id
        self.agent_id = agent_id
        self.broker_url = broker_url

        # Initialize Kafka consumer
        topics = [
            f"group.{group_id}.events",
            f"group.{group_id}.agent.{agent_id}",
        ]

        self.consumer = KafkaConsumer(
            *topics,
            bootstrap_servers=[broker_url],
            group_id=f"{agent_id}-group",
            auto_offset_reset="latest",
            value_deserializer=lambda m: json.loads(m.decode("utf-8")),
        )

        # Initialize Kafka producer for publishing events
        self.producer = KafkaProducer(
            bootstrap_servers=[broker_url],
            value_serializer=lambda v: json.dumps(v).encode("utf-8"),
            acks="all",
        )

        logger.info(
            f"Initialized Kafka consumer for {agent_id} in group {group_id}"
        )

    def _create_envelope(
        self,
        event_type: str,
        payload: Dict[str, Any],
        thread_id: Optional[str] = None,
    ) -> Dict[str, Any]:
        """
        Create a group layer envelope for publishing.

        Args:
            event_type: The event type
            payload: Event payload
            thread_id: Optional thread ID

        Returns:
            Envelope dict
        """
        if thread_id is None:
            thread_id = str(uuid4())

        return {
            "group_id": self.group_id,
            "message_id": str(uuid4()),
            "timestamp": datetime.utcnow().isoformat() + "Z",
            "thread_id": thread_id,
            "sender": {
                "id": self.agent_id,
                "role": "MEMBER",
                "agent": self.agent_id,
            },
            "type": event_type,
            "payload": payload,
            "version": "1.0",
        }

    def publish_event(
        self,
        event_type: str,
        payload: Dict[str, Any],
        thread_id: Optional[str] = None,
    ) -> str:
        """
        Publish an event to the group events topic.

        Args:
            event_type: The event type
            payload: Event payload
            thread_id: Optional thread ID

        Returns:
            Message ID
        """
        envelope = self._create_envelope(event_type, payload, thread_id)
        topic = f"group.{self.group_id}.events"

        try:
            future = self.producer.send(
                topic,
                value=envelope,
                key=envelope["thread_id"].encode("utf-8"),
            )
            future.get(timeout=10)
            logger.info(
                f"Published {event_type} event: {envelope['message_id']}"
            )
            return envelope["message_id"]
        except KafkaError as e:
            logger.error(f"Failed to publish event: {e}")
            raise GroupConsumerException(f"Failed to publish event: {e}")

    def handle_command(self, envelope: Dict[str, Any]) -> None:
        """
        Handle web fetching command.

        Args:
            envelope: The command envelope
        """
        target_agent = envelope["payload"].get("target_agent")
        command = envelope["payload"].get("command")
        args = envelope["payload"].get("args", {})
        task_id = envelope["payload"].get("task_id", str(uuid4()))
        thread_id = envelope["thread_id"]

        if target_agent != self.agent_id:
            logger.info(
                f"Command not for this agent ({target_agent} != {self.agent_id})"
            )
            return

        logger.info(f"Received command: {command} with args: {args}")

        # Publish task started event
        self.publish_event(
            "task.started",
            {
                "task_id": task_id,
                "target_agent": self.agent_id,
                "status": "started",
                "thread_id": thread_id,
            },
            thread_id,
        )

        # Import crawler service
        try:
            from crawler_service import CrawlerService
        except ImportError:
            logger.error("Failed to import crawler service")
            self.publish_event(
                "task.failed",
                {
                    "task_id": task_id,
                    "error": "Failed to import crawler service",
                    "thread_id": thread_id,
                },
                thread_id,
            )
            return

        # Execute web fetching
        try:
            url = args.get("url", "")
            if not url:
                raise ValueError("URL not provided in command args")

            logger.info(f"Fetching content from: {url}")
            crawler = CrawlerService()
            result = crawler.fetch_url(url)

            # Publish task completed event
            self.publish_event(
                "task.completed",
                {
                    "task_id": task_id,
                    "status": "completed",
                    "result": result if isinstance(result, str) else str(result),
                    "duration": 0,
                    "thread_id": thread_id,
                },
                thread_id,
            )

            # Publish result event
            self.publish_event(
                "result.web_extraction",
                {
                    "task_id": task_id,
                    "type": "web_extraction",
                    "data": {"url": url, "content": result},
                    "success": True,
                    "duration": 0,
                    "thread_id": thread_id,
                },
                thread_id,
            )

        except Exception as e:
            logger.error(f"Error executing web fetch command: {e}")
            # Publish task failed event
            self.publish_event(
                "task.failed",
                {
                    "task_id": task_id,
                    "error": str(e),
                    "status": "failed",
                    "thread_id": thread_id,
                },
                thread_id,
            )

    def process_message(self, envelope: Dict[str, Any]) -> None:
        """
        Process an incoming Kafka message.

        Args:
            envelope: The message envelope
        """
        if not envelope:
            return

        event_type = envelope.get("type", "")

        logger.info(f"Processing event: {event_type}")

        # Route by event type
        if event_type == "command.run":
            self.handle_command(envelope)
        else:
            logger.debug(f"Ignoring event type: {event_type}")

    def start(self) -> None:
        """
        Start consuming messages from Kafka.
        """
        logger.info(
            f"Starting Kafka consumer for {self.agent_id} "
            f"in group {self.group_id}"
        )

        try:
            for message in self.consumer:
                try:
                    envelope = message.value
                    self.process_message(envelope)
                except Exception as e:
                    logger.error(f"Error processing message: {e}")
                    continue

        except KeyboardInterrupt:
            logger.info("Received interrupt signal, shutting down...")
        finally:
            self.close()

    def close(self) -> None:
        """
        Close the Kafka consumer and producer.
        """
        logger.info("Closing Kafka consumer and producer...")
        try:
            self.consumer.close()
            self.producer.close()
        except Exception as e:
            logger.error(f"Error closing Kafka connections: {e}")


def main():
    """Main entry point for the Kafka consumer."""
    # Get configuration from environment
    group_id = os.getenv("GROUP_ID", "jarvis-main")
    agent_id = os.getenv("AGENT_ID", "web-fetcher")
    kafka_brokers = os.getenv("KAFKA_BROKERS", "localhost:9092")

    logger.info(
        f"Starting Web Fetcher Kafka consumer"
    )
    logger.info(f"Group ID: {group_id}")
    logger.info(f"Agent ID: {agent_id}")
    logger.info(f"Kafka Brokers: {kafka_brokers}")

    try:
        consumer = WebFetcherKafkaConsumer(
            group_id=group_id,
            agent_id=agent_id,
            broker_url=kafka_brokers,
        )
        consumer.start()
    except Exception as e:
        logger.error(f"Failed to start consumer: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()
